pymtl3/dsl/test/Placeholder_test.py

```diff
@@ -0,0 +1,299 @@
+    s.out = OutPort( mk_bits(nbits) )
+    s.w   = Wire( mk_bits(nbits) )
+    s.connect( s.w, s.out )
+
+    s.inner = Foo( 32 )( in_ = s.in_, out = s.w )
+
+  def line_trace( s ):
+    return s.inner.line_trace()
+
+def test_dumb_foo():
+  foo = Foo( 32 )
+  foo.elaborate()
+  # should elaborate just fine ..
+
+def test_dumb_foo_cannot_simulate():
+  foo = Foo( 32 )
+  foo.elaborate()
+  # should elaborate just fine ..
+  try:
+    simple_sim_pass( foo )
+  except LeftoverPlaceholderError as e:
+    print("{} is thrown\n{}".format( e.__class__.__name__, e ))
+    return
+  raise Exception("Should've thrown LeftoverPlaceholderError.")
+
+def test_foo_as_writer():
+  foo_wrap = Foo_wrap( 32 )
+
+  foo_wrap.elaborate()
+  try:
+    simple_sim_pass( foo_wrap )
+  except LeftoverPlaceholderError as e:
+    print("{} is thrown\n{}".format( e.__class__.__name__, e ))
+    return
+  raise Exception("Should've thrown LeftoverPlaceholderError.")
+
+def test_foo_field_as_writer():
+
+  class FooStruct_wrap( Component ):
+    def construct( s ):
+      s.in_ = InPort ( Bits16 )
+      s.out = OutPort( Bits32 )
+
+      s.inner = FooStruct( 16 )( in_ = s.in_ )
+      s.connect( s.inner.out.b, s.out )
+
+    def line_trace( s ):
+      return s.inner.line_trace()
+
+  foo_wrap = FooStruct_wrap()
+  foo_wrap.elaborate()
+  try:
+    simple_sim_pass( foo_wrap )
+  except LeftoverPlaceholderError as e:
+    print("{} is thrown\n{}".format( e.__class__.__name__, e ))
+    return
+  raise Exception("Should've thrown LeftoverPlaceholderError.")
+
+def test_foo_replaced_by_real():
+  foo_wrap = Foo_wrap( 32 )
+
+  foo_wrap.elaborate()
+  foo_wrap.replace_component( foo_wrap.inner, Real, check=True )
+
+  simple_sim_pass( foo_wrap )
+  foo_wrap.sim_reset()
+
+  foo_wrap.in_ = Bits32(100)
+  foo_wrap.tick()
+  print(foo_wrap.line_trace())
+  assert foo_wrap.out == 200
+
+  foo_wrap.in_ = Bits32(3)
+  foo_wrap.tick()
+  print(foo_wrap.line_trace())
+  assert foo_wrap.out == 6
+
+
+def test_cl_methodport_placeholder():
+
+  class FooCL( Placeholder, Component ):
+    def construct( s ):
+      s.enq = CalleePort()
+      s.deq = CalleePort()
+
+  class FooCL_wrap( Component ):
+    def construct( s ):
+      s.enq = CalleePort()
+      s.deq = CalleePort()
+      s.foo = FooCL()( enq = s.enq, deq = s.deq )
+
+  class FooCL_top( Component ):
+    def construct( s ):
+      s.x = FooCL_wrap()
+
+      s.counter = 0
+      @s.update
+      def up_enq():
+        s.x.enq(s.counter + 1)
+        s.counter += 1
+
+      @s.update
+      def up_recv():
+        print(s.x.deq())
+
+  class RealQueue( Component ):
+    @method_port
+    def enq( s, msg ):
+      s.q.appendleft( msg )
+    @method_port
+    def deq( s ):
+      return s.q.pop()
+
+    def construct( s ):
+      s.q = deque()
+      s.add_constraints( M(s.enq) < M(s.deq) )
+
+  a = FooCL_top()
+  a.elaborate()
+
+  a.replace_component( a.x.foo, RealQueue )
+
+  simple_sim_pass( a )
```

Comment:
Is it possible to have some preliminary assertions here? Just to make sure this works as intended.

Reply:
hmm will do